### PR TITLE
ETFE 3489 - Build view for IE704 for "Delivery section" (place of destination) in error

### PR DESCRIPTION
**Pull request by machinesleet**

### CODE REVIEW

File: test/controllers/sections/destination/DestinationIndexControllerSpec.scala [hunk 1]
```diff
@@ -60,6 +61,19 @@ class DestinationIndexControllerSpec extends SpecBase with MockUserAnswersServic
       }
     }
 
+    "when DestinationSection is not completed, and DestinationSection.status == UpdateNeeded" - {
```
Comment (LewisMitchell) [inline test/controllers/sections/destination/DestinationIndexControllerSpec.scala:64]:
```suggestion
    "when DestinationSection.status == UpdateNeeded" - {
```

File: app/forms/sections/destination/DestinationWarehouseExciseFormProvider.scala [hunk 2]
```diff
@@ -18,16 +18,23 @@ package forms.sections.destination
...
-  def apply(): Form[String] =
+
+  def apply()(implicit dataRequest: DataRequest[_]): Form[String] = {
+    val optOriginalValueSentInPreviousSubmission = DestinationWarehouseExcisePage.getOriginalAttributeValue
+
     Form(
       "value" -> text("destinationWarehouseExcise.error.required")
         .verifying(regexpUnlessEmpty(XSS_REGEX, "destinationWarehouseExcise.error.invalidCharacter"))
         .verifying(maxLength(16, "destinationWarehouseExcise.error.length"))
+        .verifying(isNotEqualToOptExistingAnswer(optOriginalValueSentInPreviousSubmission, "destinationWarehouseExcise.updateOriginalSubmission"))
```
Comment (LewisMitchell) [inline app/forms/sections/destination/DestinationWarehouseExciseFormProvider.scala:37]:
```suggestion
        .verifying(isNotEqualToOptExistingAnswer(optOriginalValueSentInPreviousSubmission, "destinationWarehouseExcise.error.submissionError"))
```

File: app/utils/SubmissionFailureErrorCodes.scala [hunk 3]
```diff
@@ -254,6 +254,33 @@ object ItemExciseProductCodeDispatchPlaceNotAllowedError {
...
+case object ExciseIdForTaxWarehouseOfDestinationNeedsConsigneeError extends SubmissionError {
+  override val code = "4412"
+  override val messageKey = "errors.704.exciseIdForTaxWarehouseOfDestinationNeedsConsignee"
+  override val id = "excise-id-for-tax-warehouse-of-destination-needs-consignee-error"
+
+  override def route()(implicit request: DataRequest[_]): Call =
+    controllers.sections.destination.routes.DestinationWarehouseExciseController.onPageLoad(request.ern, request.draftId, CheckMode)
+}
+
+case object ExciseIdForTaxWarehouseInvalid extends SubmissionError {
```
Comment (LewisMitchell) [inline app/utils/SubmissionFailureErrorCodes.scala:275]:
```suggestion
case object PlaceOfDestinationExciseIdForTaxWarehouseInvalidError extends SubmissionError {
```

File: app/utils/SubmissionFailureErrorCodes.scala [hunk 4]
```diff
@@ -254,6 +254,33 @@ object ItemExciseProductCodeDispatchPlaceNotAllowedError {
...
+case object ExciseIdForTaxWarehouseOfDestinationInvalidError extends SubmissionError {
+  override val code = "4406"
+  override val messageKey = "errors.704.exciseIdForTaxWarehouseOfDestinationInvalid"
+  override val id = "excise-id-for-tax-warehouse-of-destination-invalid-error"
+
+  override def route()(implicit request: DataRequest[_]): Call =
+    controllers.sections.destination.routes.DestinationWarehouseExciseController.onPageLoad(request.ern, request.draftId, CheckMode)
+}
+
+case object ExciseIdForTaxWarehouseOfDestinationNeedsConsigneeError extends SubmissionError {
```
Comment (LewisMitchell) [inline app/utils/SubmissionFailureErrorCodes.scala:266]:
```suggestion
case object PlaceOfDestinationNoLinkBetweenConsigneeAndPlaceOfDeliveryError extends SubmissionError {
```

File: app/utils/SubmissionFailureErrorCodes.scala [hunk 5]
```diff
@@ -254,6 +254,33 @@ object ItemExciseProductCodeDispatchPlaceNotAllowedError {
   val code = "4527"
 }
 
+case object ExciseIdForTaxWarehouseOfDestinationInvalidError extends SubmissionError {
```
Comment (LewisMitchell) [inline app/utils/SubmissionFailureErrorCodes.scala:257]:
```suggestion
case object PlaceOfDestinationExciseIdInvalidError  extends SubmissionError {
```

File: test-utils/fixtures/MovementSubmissionFailureFixtures.scala [hunk 6]
```diff
@@ -45,6 +45,14 @@ trait MovementSubmissionFailureFixtures extends BaseFixtures {
     hasBeenFixed = false
   )
 
+  val destinationWarehouseExciseFailure: MovementSubmissionFailure = MovementSubmissionFailure(
+    errorType = ExciseIdForTaxWarehouseOfDestinationInvalidError.code,
+    errorReason = "The excise ID for the tax warehouse of destination is not valid",
+    errorLocation = Some("/IE815[1]/Body[1]/SubmittedDraftOfEADESAD[1]/EadEsadDraft[1]/LocalReferenceNumber[1]"),
```
Comment (LewisMitchell) [inline test-utils/fixtures/MovementSubmissionFailureFixtures.scala:51]: Probably worth making this None in places where it's not needed (i.e., all non-item errors)

File: test-utils/fixtures/MovementSubmissionFailureFixtures.scala [hunk 7]
```diff
@@ -45,6 +45,14 @@ trait MovementSubmissionFailureFixtures extends BaseFixtures {
     hasBeenFixed = false
   )
 
+  val destinationWarehouseExciseFailure: MovementSubmissionFailure = MovementSubmissionFailure(
+    errorType = ExciseIdForTaxWarehouseOfDestinationInvalidError.code,
+    errorReason = "The excise ID for the tax warehouse of destination is not valid",
```
Comment (LewisMitchell) [inline test-utils/fixtures/MovementSubmissionFailureFixtures.scala:50]:
```suggestion
    errorReason = "Invalid or missing Place of Delivery on SEED",
```

File: test/viewmodels/helpers/CheckAnswersConsigneeHelperSpec.scala [hunk 8]
```diff
@@ -176,6 +183,17 @@ class CheckAnswersConsigneeHelperSpec extends SpecBase with MovementSubmissionFa
...
+      "the Consignee ERN needs to be updated" in new Setup(testGbWarehouseErn, ernNeedsUpdatingUserAnswers) {
+
+        val expectedSummaryListRows: Seq[SummaryListRow] = Seq(
+          ConsigneeBusinessNameSummary.row(true)(fakeDataRequest, msgs),
+          consigneeExciseSummary.row(true)(fakeDataRequest, msgs),
+          ConsigneeAddressSummary.row(true)(fakeDataRequest, msgs)
+        ).flatten
+
+        checkAnswersConsigneeHelper.summaryList() mustBe SummaryList(rows = expectedSummaryListRows)
+      }
```
Comment (LewisMitchell) [inline test/viewmodels/helpers/CheckAnswersConsigneeHelperSpec.scala:195]: I left a comment about this on Ladon's PR, I'm not sure on the value of this test. It doesn't distinguish visibly from other tests (and the summary list logic has been tested in its own spec)

File: test/views/sections/destination/DestinationWarehouseExciseViewSpec.scala [hunk 9]
```diff
@@ -17,48 +17,109 @@
...
+    "when ERN needs updating and there is NO form error" - {
+
+      implicit val msgs: Messages = messages(Seq(English.lang))
+      implicit val request: DataRequest[AnyContentAsEmpty.type] = dataRequest(FakeRequest(),
+        emptyUserAnswers.copy(submissionFailures =
+          DestinationWarehouseExcisePage.possibleErrors.map(error => destinationWarehouseExciseFailure.copy(error.code))
+        )
+      )
+
+      lazy val view = app.injector.instanceOf[DestinationWarehouseExciseView]
```
Comment (LewisMitchell) [inline test/views/sections/destination/DestinationWarehouseExciseViewSpec.scala:80]: Can this be instantiated within the class scope?

File: app/views/sections/destination/DestinationCheckAnswersView.scala.html
Comment (LewisMitchell) [inline app/views/sections/destination/DestinationCheckAnswersView.scala.html:1]: I think this needs tests

File: app/views/sections/destination/DestinationWarehouseExciseView.scala.html [hunk 10]
```diff
@@ -38,6 +41,10 @@
 
         @if(form.errors.nonEmpty) {
             @govukErrorSummary(ErrorSummaryViewModel(form))
+        } else {
+            @errorNotificationBannerHelper.content(
+                errors = DestinationWarehouseExcisePage.getMovementSubmissionErrors
+            ).map(notificationBanner(_))
```
Comment (LewisMitchell) [inline app/views/sections/destination/DestinationWarehouseExciseView.scala.html:47]: Currently looks like this:
image

```suggestion
            @errorNotificationBannerHelper.content(
                errors = DestinationWarehouseExcisePage.getMovementSubmissionErrors,
                hasContentHeading = false
            ).map(notificationBanner(_))
```

**Review by LewisMitchell — CHANGES_REQUESTED** [review]

File: test/controllers/sections/destination/DestinationIndexControllerSpec.scala (hunk 1 above)
Comment (machinesleet) [reply test/controllers/sections/destination/DestinationIndexControllerSpec.scala:64]: I wanted to be explicit about this, as the DestinationSection not being completed is the only way the second part of the or condition is evaluated

File: test/viewmodels/helpers/CheckAnswersConsigneeHelperSpec.scala (hunk 8 above)
Comment (machinesleet) [reply test/viewmodels/helpers/CheckAnswersConsigneeHelperSpec.scala:195]: Can't remember changing this one, I think I've got some stuff leftover from Ladon's branch - will remove

**Review by LewisMitchell — APPROVED** [review]

**Pull request merged**